src/ray/core_worker/store_provider/plasma_store_provider.cc

```diff
@@ -83,10 +83,11 @@ Status CoreWorkerPlasmaStoreProvider::Seal(const ObjectID &object_id) {
 
 Status CoreWorkerPlasmaStoreProvider::FetchAndGetFromPlasmaStore(
     absl::flat_hash_set<ObjectID> &remaining, const std::vector<ObjectID> &batch_ids,
-    int64_t timeout_ms, bool fetch_only, const TaskID &task_id,
+    int64_t timeout_ms, bool fetch_only, bool in_direct_call, const TaskID &task_id,
     absl::flat_hash_map<ObjectID, std::shared_ptr<RayObject>> *results,
     bool *got_exception) {
-  RAY_RETURN_NOT_OK(raylet_client_->FetchOrReconstruct(batch_ids, fetch_only, task_id));
+  RAY_RETURN_NOT_OK(raylet_client_->FetchOrReconstruct(
+      batch_ids, fetch_only, /*mark_worker_blocked*/ !in_direct_call, task_id));
```

Comment:
I think we could avoid making the changes in node_manager.cc if we just set the `fetch_only` flag correctly: something like `in_direct_call ? true : fetch_only`. Maybe less explicit but this was also a bit confusing to me.

Reply:
I'm not sure I follow. Isn't fetch only separate?

Reply:
The new flag seems to do exactly the same thing as fetch_only=true, so I was suggesting we just reuse the fetch_only flag. So instead of passing both fetch_only and mark_worker_blocked as fields into FetchOrReconstruct, we'd just pass `in_direct_call ? true : fetch_only`.

Reply:
After looking into this flag more I'm not even sure this PR is correct currently now, let's follow up offline.